x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/history/WatchRecord.java

```diff
@@ -356,5 +418,104 @@ void innerToXContent(XContentBuilder builder, Params params) throws IOException
+     * drops other information to avoid having a document that is too large to index into Elasticsearch.
+     */
+    private static ExecutableInput<? extends Input, ? extends Input.Result> getTruncatedInput() {
+        return new ExecutableInput<>(new Input() {
+            @Override
+            public String type() {
+                return TRUNCATED_RECORD_KEY;
+            }
+
+            @Override
+            public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+                builder.value(TRUNCATED_RECORD_VALUE);
+                // builder.startObject();
+                // builder.field(TRUNCATED_RECORD_KEY, TRUNCATED_RECORD_VALUE);
+                // builder.endObject();
+                return builder;
+            }
+        }) {
+            @Override
+            public Input.Result execute(WatchExecutionContext ctx, Payload payload) {
+                throw new UnsupportedOperationException("Redacted input cannot be executed");
+            }
+        };
+    }
+
+    /*
+     * This returns a WatchExecutionResult whose toXContent() returns minimal information, including a new TRUNCATED_MESSAGE field. It
+     * drops most other information to avoid having a document that is too large to index into Elasticsearch.
+     */
+    private static WatchExecutionResult getTruncatedWatchExecutionResult(WatchRecord watchRecord) {
+        WatchExecutionContext watchExecutionContext = new WatchExecutionContext(
+            watchRecord.id.watchId(),
+            watchRecord.executionResult.executionTime(),
+            null,
+            TimeValue.ZERO
+        ) {
+            @Override
+            public boolean knownWatch() {
+                return false;
+            }
+
+            @Override
+            public boolean simulateAction(String actionId) {
+                return false;
+            }
+
+            @Override
+            public boolean skipThrottling(String actionId) {
+                return false;
+            }
+
+            @Override
+            public boolean shouldBeExecuted() {
+                return false;
+            }
+
+            @Override
+            public boolean recordExecution() {
+                return false;
+            }
+        };
+
+        return new WatchExecutionResult(watchExecutionContext, watchRecord.executionResult.executionDurationMs()) {
+            @Override
+            public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+                builder.startObject();
+                WatcherDateTimeUtils.writeDate(
+                    Field.EXECUTION_TIME.getPreferredName(),
+                    builder,
+                    watchRecord.executionResult.executionTime()
+                );
+                builder.field(Field.EXECUTION_DURATION.getPreferredName(), watchRecord.executionResult.executionDurationMs());
+                builder.field(TRUNCATED_RECORD_KEY, TRUNCATED_RECORD_VALUE);
```

Comment:
I'm also assuming that adding a new field to the history store doesn't need any further changes to mappings or anything?

Reply:
It looks like `input` doesn't specify dynamic, so it defaults to dynamic (https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/template-resources/src/main/resources/watch-history.json#L108). And `result` is explicitly dynamic (https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/template-resources/src/main/resources/watch-history.json#L136).